ADAL/src/validation/ADAadAuthorityCache.m

```diff
@@ -0,0 +1,274 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "ADAadAuthorityCache.h"
+
+#import "NSURL+ADExtensions.h"
+
+#include <pthread.h>
+
+#define CHECK_CLASS_TYPE(_CHK, _CLS, _ERROR) \
+    if (![_CHK isKindOfClass:[_CLS class]]) { \
+        ADAuthenticationError *adError = \
+        [ADAuthenticationError errorFromAuthenticationError:AD_ERROR_SERVER_INVALID_RESPONSE \
+                                       protocolCode:nil \
+                                       errorDetails:_ERROR \
+                                      correlationId:context.correlationId]; \
+        if (error) { *error = adError; } \
+        return NO; \
+    }
+
+@implementation ADAadAuthorityCacheRecord
+
+@end
+
+@implementation ADAadAuthorityCache
+
+- (id)init
+{
+    if (!(self = [super init]))
+    {
+        return nil;
+    }
+    
+    
+    _recordMap = [NSMutableDictionary new];
+    pthread_rwlock_init(&_rwLock, NULL);
+    
+    return self;
+}
+
+- (void)dealloc
+{
+    pthread_rwlock_destroy(&_rwLock);
+}
+
+- (BOOL)processMetadata:(NSArray<NSDictionary *> *)metadata
+              authority:(NSURL *)authority
+                context:(id<ADRequestContext>)context
+                  error:(ADAuthenticationError * __autoreleasing *)error
+{
+    if (metadata != nil)
+    {
+        CHECK_CLASS_TYPE(metadata, NSArray, @"JSON metadata from authority validation is not an array");
+    }
+    
+    [self getWriteLock];
+    BOOL ret = [self processImpl:metadata authority:authority context:context error:error];
+    pthread_rwlock_unlock(&_rwLock);
+    
+    return ret;
+}
+
+- (BOOL)processImpl:(NSArray<NSDictionary *> *)metadata
+          authority:(NSURL *)authority
+            context:(id<ADRequestContext>)context
+              error:(ADAuthenticationError * __autoreleasing *)error
+{
+    if (metadata.count == 0)
+    {
+        AD_LOG_INFO(@"No metadata returned from authority validation", context.correlationId, nil);
+    }
+    else
+    {
+        AD_LOG_INFO(@"Caching AAD Environements:", context.correlationId, nil);
+    }
+    
+    NSMutableArray<ADAadAuthorityCacheRecord *> *recordsToAdd = [NSMutableArray new];
+    
+    for (NSDictionary *environment in metadata)
+    {
+        CHECK_CLASS_TYPE(environment, NSDictionary, @"JSON metadata entry is not a dictionary");
+        
+        __auto_type *record = [ADAadAuthorityCacheRecord new];
```

Comment:
btw, we can make it a bit fancier :), using defines it will be similar to Swift's var & let:
```
#define let __auto_type const //"let" alias for constants
#define var __auto_type //"var" alias for variables
 
var stringVar = @"Hello"; 
let stringConst = [NSString stringWithFormat:@"Hello"];
```